### PR TITLE
refactor: Remove AntaMissingParamException 

### LINKED ISSUE #405: as discussed internally need to revisit some tests to add `AntaMissingParamException`
as discussed internally need to revisit some tests to add `AntaMissingParamException`

_Originally posted by @gmuloc in https://github.com/arista-netdevops-community/anta/issues/403#issuecomment-1730074898_

### PR DESCRIPTION
**Pull request by gmuloc**
Fixes #405 

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have run pre-commit for code linting and typing (`pre-commit run`)
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes (`tox -e testenv`)

### CODE REVIEW

File: anta/tests/connectivity.py [hunk 1]
```diff
@@ -42,7 +42,12 @@ class VerifyReachability(AntaTest):
     name = "VerifyReachability"
     description = "Test the network reachability to one or many destination IP(s)."
     categories: ClassVar[list[str]] = ["connectivity"]
-    commands: ClassVar[list[AntaCommand | AntaTemplate]] = [AntaTemplate(template="ping vrf {vrf} {destination} source {source} repeat {repeat}", revision=1)]
+    commands: ClassVar[list[AntaCommand | AntaTemplate]] = [
+        AntaTemplate(
+            template="ping vrf {vrf} {destination} source {source} repeat {repeat}",
+            revision=1,
+        )
+    ]
```
Comment (carl-baillargeon) [inline anta/tests/connectivity.py:50]: Same issue as yesterday?

File: anta/models.py [hunk 2]
```diff
@@ -430,13 +447,15 @@ def failed_commands(self) -> list[AntaCommand]:
...
-    def render(self, template: AntaTemplate) -> list[AntaCommand]:  # pylint: disable=W0613  # noqa: ARG002
+    def render(self, template: AntaTemplate) -> list[AntaCommand]:
         """Render an AntaTemplate instance of this AntaTest using the provided AntaTest.Input instance at self.inputs.
 
         This is not an abstract method because it does not need to be implemented if there is
         no AntaTemplate for this test.
         """
+        # Disabling unused argument
+        # ruff: noqa: ARG002
+        # pylint: disable=W0613
```
Comment (carl-baillargeon) [inline anta/models.py:458]: We can add something like `_ = template`. This is what I did when adding ruff

File: anta/models.py [hunk 3]
```diff
@@ -40,16 +41,21 @@
...
-    This Exception should in general never be raised in normal usage of ANTA.
-    """
+    model_config = ConfigDict(extra="forbid")
 
-    def __init__(self, message: str) -> None:
-        """Append Github suggestion to message."""
-        self.message = f"{message}\n{GITHUB_SUGGESTION}"
-        super().__init__(self.message)
+    if not TYPE_CHECKING:
+        # They put __getattr__ in not TYPE_CHEKCING in the lib
```
Comment (carl-baillargeon) [inline anta/models.py:50]:
```suggestion
        # They put __getattr__ in not TYPE_CHECKING in the lib
```

Comment (carl-baillargeon) [reply anta/models.py:50]: Who is They and which lib are we talking about?

File: anta/tests/connectivity.py [hunk 4]
```diff
@@ -64,20 +69,24 @@ class Host(BaseModel):
...
-        return [template.render(destination=host.destination, source=host.source, vrf=host.vrf, repeat=host.repeat) for host in self.inputs.hosts]
+        return [
+            template.render(
+                destination=host.destination,
+                source=host.source,
+                vrf=host.vrf,
+                repeat=host.repeat,
+            )
+            for host in self.inputs.hosts
+        ]
```
Comment (carl-baillargeon) [inline anta/tests/connectivity.py:80]: Same

File: anta/tests/routing/bgp.py [hunk 5]
```diff
@@ -549,17 +606,27 @@ def test(self) -> None:
...
             # Swapping AFI and SAFI in case of SR-TE
             if afi == "sr-te":
                 afi, safi = safi, afi
-            afi_vrf = cast(str, command.params.get("vrf"))
-            afi_peers = cast(list[Union[IPv4Address, IPv6Address]], command.params.get("peers", []))
+
+            for input_entry in self.inputs.address_families:
+                if input_entry.afi == afi and input_entry.safi == safi and input_entry.vrf == afi_vrf:
+                    afi_peers = input_entry.peers
+                    break
```
Comment (carl-baillargeon) [inline anta/tests/routing/bgp.py:620]: Should we have a function for this?

File: anta/tests/routing/bgp.py [hunk 6]
```diff
@@ -650,10 +726,13 @@ def test(self) -> None:
...
-            vrf = command.params["vrf"]
-            advertised_routes = command.params["advertised_routes"]
-            received_routes = command.params["received_routes"]
+            peer = command.params.peer
+            vrf = command.params.vrf
+            for input_entry in self.inputs.bgp_peers:
+                if str(input_entry.peer_address) == peer and input_entry.vrf == vrf:
+                    advertised_routes = input_entry.advertised_routes
+                    received_routes = input_entry.received_routes
+                    break
```
Comment (carl-baillargeon) [inline anta/tests/routing/bgp.py:735]: Not sure if we should have a different way of getting the inputs (related to the other comment)

File: anta/tests/security.py [hunk 7]
```diff
@@ -621,16 +642,21 @@ class IPv4ACLEntry(BaseModel):
...
             # Collecting input ACL details
-            acl_name = command_output.params["acl"]
-            acl_entries = command_output.params["entries"]
+            acl_name = command_output.params.acl
+            # Retrieve the expected entries from the inputs
+            acl_entries = []
+            for acl in self.inputs.ipv4_access_lists:
+                if acl.name == acl_name:
+                    acl_entries = acl.entries
+                    break
```
Comment (carl-baillargeon) [inline anta/tests/security.py:659]: Related to other comments

File: anta/tests/routing/bgp.py [hunk 8]
```diff
@@ -261,20 +277,31 @@ def test(self) -> None:
...
             peer_count = 0
             command_output = command.json_output
 
-            afi = cast(Afi, command.params.get("afi"))
-            safi = cast(Optional[Safi], command.params.get("safi"))
-            afi_vrf = cast(str, command.params.get("vrf"))
-            num_peers = cast(PositiveInt, command.params.get("num_peers"))
+            afi = command.params.afi
+            safi = command.params.safi
+            afi_vrf = command.params.vrf or "default"
```
[the remark this answers is not in the document]
Comment (gmuloc) [reply anta/tests/routing/bgp.py:272]: I put it everywhere where we had this as some command are with vrf and some are not and then we were just passing vrf default blindly

File: anta/tests/routing/bgp.py (hunk 5 above)
Comment (gmuloc) [reply anta/tests/routing/bgp.py:576]: yes! maybe in another PR?

Comment (carl-baillargeon) [reply anta/tests/routing/bgp.py:576]: Sure. Kind of related to issue #388

File: anta/tests/services.py [hunk 9]
```diff
@@ -16,6 +14,9 @@
 from anta.models import AntaCommand, AntaTemplate, AntaTest
 from anta.tools import get_dict_superset, get_failed_logs, get_item
 
+# Mypy does not understand AntaTest.Input typing
+# mypy: disable-error-code=attr-defined
+
```
Comment (carl-baillargeon) [inline anta/tests/services.py:19]: Why this has been moved below?

File: anta/tests/routing/bgp.py [hunk 10]
```diff
@@ -384,8 +391,11 @@ def render(self, template: AntaTemplate) -> list[AntaCommand]:
             # For SR-TE SAFI, the EOS command supports sr-te first then ipv4/ipv6
             elif template == VerifyBGPPeersHealth.commands[0] and afi.afi in ["ipv4", "ipv6"] and afi.safi == "sr-te":
                 commands.append(template.render(afi=afi.safi, safi=afi.afi, vrf=afi.vrf))
-            elif template == VerifyBGPPeersHealth.commands[1] and afi.afi not in ["ipv4", "ipv6"]:
-                commands.append(template.render(afi=afi.afi, vrf=afi.vrf))
+            elif template == VerifyBGPPeersHealth.commands[1] and afi.afi not in [
+                "ipv4",
+                "ipv6",
+            ]:
```
Comment (carl-baillargeon) [inline anta/tests/routing/bgp.py:397]: Same line no?

File: anta/tests/routing/bgp.py [hunk 11]
```diff
@@ -244,13 +243,16 @@ def render(self, template: AntaTemplate) -> list[AntaCommand]:
...
             # For SR-TE SAFI, the EOS command supports sr-te first then ipv4/ipv6
             elif template == VerifyBGPPeerCount.commands[0] and afi.afi in ["ipv4", "ipv6"] and afi.safi == "sr-te":
-                commands.append(template.render(afi=afi.safi, safi=afi.afi, vrf=afi.vrf, num_peers=afi.num_peers))
-            elif template == VerifyBGPPeerCount.commands[1] and afi.afi not in ["ipv4", "ipv6"]:
-                commands.append(template.render(afi=afi.afi, vrf=afi.vrf, num_peers=afi.num_peers))
+                commands.append(template.render(afi=afi.safi, safi=afi.afi, vrf=afi.vrf))
+            elif template == VerifyBGPPeerCount.commands[1] and afi.afi not in [
+                "ipv4",
+                "ipv6",
+            ]:
```
Comment (carl-baillargeon) [inline anta/tests/routing/bgp.py:254]: Same line?

File: anta/tests/routing/bgp.py [hunk 12]
```diff
@@ -142,7 +138,10 @@ def _add_bgp_routes_failure(
 
         # If the route is either inactive or invalid, add it to the failure routes dictionary
         if not is_active or not is_valid:
-            failure["bgp_peers"][peer][vrf][route_type][str_route] = {"valid": is_valid, "active": is_active}
+            failure["bgp_peers"][peer][vrf][route_type][str_route] = {
+                "valid": is_valid,
+                "active": is_active,
+            }
```
Comment (carl-baillargeon) [inline anta/tests/routing/bgp.py:144]: :crying_cat_face:

File: anta/tests/security.py [hunk 13]
```diff
@@ -621,16 +622,21 @@ class IPv4ACLEntry(BaseModel):
...
         self.result.is_success()
         for command_output in self.instance_commands:
             # Collecting input ACL details
-            acl_name = command_output.params["acl"]
-            acl_entries = command_output.params["entries"]
+            acl_name = command_output.params.acl
+            # Retrieve the expected entries from the inputs
+            acl_entries = []
+            for acl in self.inputs.ipv4_access_lists:
+                if acl.name == acl_name:
```
Comment (MaheshGSLAB) [inline anta/tests/security.py:637]:
```
for command_output, acl in zip(self.instance_commands, self.inputs.ipv4_access_lists):
  # Collecting input ACL details
  acl_name = command_output.params.acl
  # Retrieve the expected entries from the inputs
  acl_entries = acl.entries
```

**Pull request merged**